### PR TITLE
Introduce user management model

### PR DESCRIPTION
**Pull request by sven1103**

### CODE REVIEW

File: domain/src/main/java/life/qbic/usermanagement/User.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+   * QBiC's organisation.
+   *
+   * @param password the desired password for the user
+   * @param fullName the full name of the user
+   * @param email    the email address of the user
+   * @return the new user
+   * @since 1.0.0
+   */
+  public static User create(String password, String fullName, String email) throws UserException {
+    String uuid = String.valueOf(UUID.randomUUID());
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/User.java:27]: do we need to check if this id is already in use?

File: domain/src/main/java/life/qbic/usermanagement/policies/EmailFormatPolicy.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * <p>Validates a String against the RFC5322 email format specification (<a href="https://www.rfc-editor.org/rfc/rfc5322">https://www.rfc-editor.org/rfc/rfc5322</a>).</p>
+ *
+ * @since 1.0.0
+ */
+public class EmailFormatPolicy {
+
+  /*
+  Many thanks to https://www.javatpoint.com/java-email-validation
+   */
+  private static final String FULL_ADDRESS_SPEC = "^[\\w!#$%&'*+/=?`{|}~^-]+(?:\\.[\\w!#$%&'*+/=?`{|}~^-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,6}$";
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/policies/EmailFormatPolicy.java:17]: 👀

File: domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    this.dataStorage = dataStorage;
+  }
+
+  public Optional<User> findByEmail(String email) {
+    var matchingUsers = dataStorage.findUsersByEmail(email);
+    if (matchingUsers.size() > 1) {
+      throw new RuntimeException("More than one user entry with the same email exists!");
+    }
+    return Optional.empty();
+  }
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java:28]: You actually never return a user here, right?

File: domain/src/test/groovy/life/qbic/usermanagement/UserSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    def "When a new user is created, a unique identifier is assigned to the user"() {
+        when:
+        User user = User.create("test1234", "My Name", "my.name@example.com")
+
+        then:
+        !generatedUserIds.contains(user.getId())
+        generatedUserIds.add(user.getId())
+
+        where:
+        run << (1..10_000)
```
Comment (jenniferboedker) [inline domain/src/test/groovy/life/qbic/usermanagement/UserSpec.groovy:29]: what does "run" mean here?

File: domain/src/test/groovy/life/qbic/usermanagement/UserSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    def "When an invalid email is provided, throw a user exception"() {
+        when:
+        User.create("123", "My Name", "my.name@example.com")
```
Comment (jenniferboedker) [inline domain/src/test/groovy/life/qbic/usermanagement/UserSpec.groovy:34]: I would use here a correct password (of length 8) to only test for a valid email.
Also, the email is valid, right?

File: domain/src/test/groovy/life/qbic/usermanagement/UserSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import spock.lang.Specification
+import spock.lang.Unroll
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline domain/src/test/groovy/life/qbic/usermanagement/UserSpec.groovy:13]: missing description

File: domain/src/test/groovy/life/qbic/usermanagement/policies/PasswordPolicySpec.groovy [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+import spock.lang.Specification
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline domain/src/test/groovy/life/qbic/usermanagement/policies/PasswordPolicySpec.groovy:11]: description missing

File: domain/src/test/groovy/life/qbic/usermanagement/repository/UserRepositorySpec.groovy [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import life.qbic.usermanagement.User
+import spock.lang.Specification
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline domain/src/test/groovy/life/qbic/usermanagement/repository/UserRepositorySpec.groovy:12]: description missing

File: domain/src/test/groovy/life/qbic/usermanagement/repository/UserRepositorySpec.groovy [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        UserRepository repository = new UserRepository(storage)
+
+        when:
+        repository.findByEmail("my.example@example.com")
+
+        then:
+        thrown(RuntimeException)
+    }
+
+    def "Given a repository already contains a user, dont add the user twice"() {
```
Comment (jenniferboedker) [inline domain/src/test/groovy/life/qbic/usermanagement/repository/UserRepositorySpec.groovy:29]: This is not really what you test, right?
You test if a user is already stored. This test does not cover if the person is created again.

**Review by jenniferboedker — COMMENTED** [review]
Overall looks very good, I like that the policies and internal rules are so explicit!
Left some comments where things were unclear :)

File: domain/src/main/java/life/qbic/usermanagement/User.java (hunk 1 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/usermanagement/User.java:27]: a) Covered by tests
b) UUID version 4 collision probability is low:
https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)

File: domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java (hunk 3 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java:28]: ha, nice spot!

**Review by jenniferboedker — DISMISSED** [review]
🚀

File: domain/src/main/java/life/qbic/usermanagement/User.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
+package life.qbic.usermanagement;
+
+import java.util.UUID;
+import jdk.jshell.spi.ExecutionControl.UserException;
+import life.qbic.usermanagement.policies.*;
+
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/User.java:6]: missing java doc here

File: domain/src/main/java/life/qbic/usermanagement/policies/PasswordEncryptionPolicy.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package life.qbic.usermanagement.policies;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/policies/PasswordEncryptionPolicy.java:5]: Missing Javadocs in this class ;)

File: domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
+package life.qbic.usermanagement.repository;
+
+import java.util.Optional;
+import life.qbic.usermanagement.User;
+
+/**
+ * Provides stateless access and storage functionality for {@link User} entities.
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java:7]:
```suggestion
 * <b> Provides stateless access and storage functionality for {@link User} entities. </b>
```

File: domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+ */
+public class UserRepository {
+
+  private static UserRepository INSTANCE;
+
+  private final UserDataStorage dataStorage;
+
+  /**
+   * Retrieves a Singleton instance of a user {@link UserRepository}. In case this method is called
+   * the first time, a new instance is created one.
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/repository/UserRepository.java:19]:
```suggestion
   * the first time, a new instance is created.
```

File: domain/src/main/java/life/qbic/usermanagement/policies/PasswordEncryptionPolicy.java (hunk 11 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/usermanagement/policies/PasswordEncryptionPolicy.java:5]: added

File: domain/src/main/java/life/qbic/usermanagement/policies/PasswordEncryptionPolicy.java [hunk 14]
```diff
@@ -3,16 +3,41 @@
...
+ *
+ * @since 1.0.0
+ */
 public class PasswordEncryptionPolicy {
+
   private static PasswordEncryptionPolicy INSTANCE;
 
+  /**
+   * Creates a {@link PasswordEncryptionPolicy} reference. Following the singleton pattern an object
+   * is only instanced once, when this method is called the first time.
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/policies/PasswordEncryptionPolicy.java:20]:
```suggestion
   * is only instanced once this method is called the first time.
```

**Review by jenniferboedker — APPROVED** [review]

**Pull request merged**